File: python/11_encapsulation.py
# ============================================================
# Topic:  Encapsulation
# Summary: Hide state behind methods; validate changes; expose read-only views.
# Run:     python python/11_encapsulation.py
# ============================================================

class BankAccount:
    def __init__(self, owner, balance=0):
        self.owner = owner
        self.__balance = balance      # __ = "private" (name-mangled to _BankAccount__balance)

    # ----- read-only accessor via @property (no setter = can't assign) -----
    @property
    def balance(self):
        return self.__balance

    # ----- controlled mutation: validate before changing state -----
    def deposit(self, amount):
        if amount > 0:
            self.__balance += amount

    def withdraw(self, amount):
        if amount > self.__balance:
            return False              # reject: not enough money
        self.__balance -= amount
        return True


acct = BankAccount("Vetri", 100)
acct.deposit(50)
print("Balance:", acct.balance)       # Output: Balance: 150
acct.withdraw(30)
print("Balance:", acct.balance)       # Output: Balance: 120
if not acct.withdraw(1000):
    print("Insufficient funds")       # Output: Insufficient funds
print("Balance:", acct.balance)       # Output: Balance: 120
# acct.balance = 999  # AttributeError: property 'balance' has no setter

# Cross-language note:
#   Python has no real access control — `__name` only name-mangles by convention.
#   @property exposes a read-only view; the invariant (balance >= 0) lives in the methods.
#   Other languages enforce privacy with keywords (private) checked by the compiler.

File: csharp/11_encapsulation.cs
// ============================================================
// Topic:  Encapsulation
// Summary: Hide state behind methods; validate changes; expose read-only views.
// Run:     dotnet script csharp/11_encapsulation.cs
// ============================================================

using System;

var acct = new BankAccount("Vetri", 100);
acct.Deposit(50);
Console.WriteLine($"Balance: {acct.Balance}"); // Output: Balance: 150
acct.Withdraw(30);
Console.WriteLine($"Balance: {acct.Balance}"); // Output: Balance: 120
if (!acct.Withdraw(1000))
{
    Console.WriteLine("Insufficient funds");    // Output: Insufficient funds
}
Console.WriteLine($"Balance: {acct.Balance}"); // Output: Balance: 120
// acct.Balance = 999; // compile error: the set accessor is private

class BankAccount
{
    // auto-property with a PRIVATE setter: readable outside, writable only inside
    public string Owner { get; }
    public int Balance { get; private set; }

    public BankAccount(string owner, int balance)
    {
        Owner = owner;
        Balance = balance;
    }

    // ----- controlled mutation: validate before changing state -----
    public void Deposit(int amount)
    {
        if (amount > 0) Balance += amount;
    }

    public bool Withdraw(int amount)
    {
        if (amount > Balance) return false; // reject
        Balance -= amount;
        return true;
    }
}

// Cross-language note:
//   C# properties replace getter/setter boilerplate: `{ get; private set; }` is a
//   read-only-from-outside view with internal mutation. `private` is compiler-enforced.
//   Validation lives in Deposit/Withdraw so the balance invariant can't be bypassed.
